### ISSUE TITLE
How to use Gemini with txtai

### ISSUE THREAD

**Issue by igorlima** [neuml/txtai#843]
I've been exploring the possibilities of using Google Gemini with `txtai`, but I haven't found any references to Gemini in the documentation yet. 

Is there a way to embed text in `txtai` using Gemini? The documentation [references other LLMs](https://neuml.github.io/txtai/pipeline/text/llm/), but Gemini seems missing.

Here's a snippet of what I've attempted using the [`litellm` method](https://docs.litellm.ai/docs/providers/gemini), though I haven't had any success so far:

```python
import os
os.environ["GEMINI_API_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"

from txtai import LLM
llm = LLM("gemini/gemini-pro", method="litellm")
llm("Where is one place you'd go in Washington, DC?")

from txtai import Embeddings
embeddings = Embeddings(path="litellm/gemini/gemini-pro")
data = [
  "US tops 5 million confirmed virus cases",
  "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
]
embeddings.index(data)
```

I'd [appreciate the guidance](https://medium.com/@igorribeirolima/i-just-wanted-to-say-how-much-i-appreciate-your-contributions-to-the-txtai-community-3eecf1f91dbe) if anyone has insights or knows of any documentation or examples on using Gemini with `txtai`.

**Comment by davidmezzetti** [comment]
Hello, first off thank you for the kind words!

What kind of error are you receiving?

**Comment by igorlima** [comment]
Before jumping into the error I'm facing, let me share my setup. I use a Docker container running Python 3.12.2 on MacOS.

- <details><summary>here's a quick peek at my setup:</summary>
  
  - I start the Docker container with this command:
    ```sh
    docker run \
      --name python-learning --rm \
      --mount src=`realpath .`,target=/home/local,type=bind \
      --workdir /home/local \
      -it python:3.12.2 bash
    ```
  - Since the Docker container is Linux-based, I needed to install `libsndfile`. You can find more info [here](https://neuml.github.io/txtai/install/#linux) and [here](https://github.com/libsndfile/libsndfile/issues/297#issuecomment-312946414). Here's a handy command to get it installed:
    ```docker-container
    apt update
    apt install libsndfile1
    ```
  </details>

- <details><summary>a Python script (<code>main.py</code>):</summary>
  
  I've been playing around with the following Python script (`main.py`) to test the Gemini LLM model. Here's a snippet:
  ```python
  import os
  os.environ["GEMINI_API_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
  
  from txtai import LLM
  llm = LLM("gemini/gemini-pro", method="litellm")
  llm("Where is one place you'd go in Washington, DC?")
  
  from txtai import Embeddings
  embeddings = Embeddings(path="litellm/gemini/gemini-pro")
  
  data = [
    "US tops 5 million confirmed virus cases",
    "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
    "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
    "The National Park Service warns against sacrificing slower friends in a bear attack",
    "Maine man wins $1M from $25 lottery ticket",
    "Make huge profits without work, earn up to $100,000 a day"
  ]
  
  embeddings.index(data)
  print("%-20s %s" % ("Query", "Best Match"))
  print("-" * 50)
  for query in ("feel good story", "climate change", "public health story", "war", "wildlife", "asia", "lucky", "dishonest junk"):
    uid = embeddings.search(query, 1)[0][0]
    print("%-20s %s" % (query, data[uid]))
  ```
  </details>


After setting everything up and running `python3 main.py`, I ran into this error:
- <details><summary>the entire log error</summary>
  
  ```txt
  Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
  LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.
  
  Traceback (most recent call last):
    File "/usr/local/lib/python3.12/site-packages/litellm/main.py", line 2262, in completion
      response = vertex_chat_completion.completion(  # type: ignore
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/vertex_and_google_ai_studio_gemini.py", line 1228, in completion
      data = sync_transform_request_body(**transform_request_params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 379, in sync_transform_request_body
      return _transform_request_body(
             ^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 342, in _transform_request_body
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 307, in _transform_request_body
      content = litellm.GoogleAIStudioGeminiConfig()._transform_messages(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/gemini/chat/transformation.py", line 131, in _transform_messages
      return _gemini_convert_messages_with_history(messages=messages)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 262, in _gemini_convert_messages_with_history
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 253, in _gemini_convert_messages_with_history
      raise Exception(
  Exception: Invalid Message passed in - {'content': "Where is one place you'd go in Washington, DC?", 'role': 'prompt'}. File an issue https://github.com/BerriAI/litellm/issues
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/home/local/python/2024a11m24d-txtai-embeddings/main-gemini.py", line 32, in <module>
      llm("Where is one place you'd go in Washington, DC?")
    File "/usr/local/lib/python3.12/site-packages/txtai/pipeline/llm/llm.py", line 63, in __call__
      return self.generator(text, maxlength, stream, stop, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/txtai/pipeline/llm/generation.py", line 54, in __call__
      results = self.execute(texts, maxlength, stream, stop, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/txtai/pipeline/llm/generation.py", line 86, in execute
      return list(self.stream(texts, maxlength, stream, stop, **kwargs))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/txtai/pipeline/llm/litellm.py", line 60, in stream
      result = api.completion(
               ^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/utils.py", line 987, in wrapper
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/utils.py", line 868, in wrapper
      result = original_function(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/main.py", line 3012, in completion
      raise exception_type(
            ^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/litellm_core_utils/exception_mapping_utils.py", line 2146, in exception_type
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/litellm_core_utils/exception_mapping_utils.py", line 2122, in exception_type
      raise APIConnectionError(
  litellm.exceptions.APIConnectionError: litellm.APIConnectionError: Invalid Message passed in - {'content': "Where is one place you'd go in Washington, DC?", 'role': 'prompt'}. File an issue https://github.com/BerriAI/litellm/issues
  Traceback (most recent call last):
    File "/usr/local/lib/python3.12/site-packages/litellm/main.py", line 2262, in completion
      response = vertex_chat_completion.completion(  # type: ignore
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/vertex_and_google_ai_studio_gemini.py", line 1228, in completion
      data = sync_transform_request_body(**transform_request_params)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 379, in sync_transform_request_body
      return _transform_request_body(
             ^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 342, in _transform_request_body
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 307, in _transform_request_body
      content = litellm.GoogleAIStudioGeminiConfig()._transform_messages(
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/gemini/chat/transformation.py", line 131, in _transform_messages
      return _gemini_convert_messages_with_history(messages=messages)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 262, in _gemini_convert_messages_with_history
      raise e
    File "/usr/local/lib/python3.12/site-packages/litellm/llms/vertex_ai/gemini/transformation.py", line 253, in _gemini_convert_messages_with_history
      raise Exception(
  Exception: Invalid Message passed in - {'content': "Where is one place you'd go in Washington, DC?", 'role': 'prompt'}. File an issue https://github.com/BerriAI/litellm/issues
  ```
  </details>
- | image A | image B |
  | :---: | :---: |
  | ![image](https://github.com/user-attachments/assets/799b8404-ba1e-4a78-882d-b5a3ad6cb9eb) | ![image](https://github.com/user-attachments/assets/fd0f456b-f102-4fa4-9b1d-555de64a75e6) |

**In short**, it outputs the following error message:
```txt
Exception: Invalid Message passed in - {'content': "Where is one place you'd go in Washington, DC?", 'role': 'prompt'}. File an issue https://github.com/BerriAI/litellm/issues
```

**Comment by davidmezzetti** [comment]
Looks like you're running into the same issue as this: https://github.com/neuml/txtai/issues/841

The original LLM pipeline was designed to work with raw prompts (i.e. with chat templates manually applied). Over time this has changed where chat messages are more the default. The issue above added a new `defaultrole` parameter to address this. At some point soon, the switch will be flipped to make the `defaultrole="user"` and issues like this will go away. 

Nonetheless, here are the options in the meantime.

1. Pass chat messages instead of strings to the LLM pipeline.

```python
llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}])
```

2. Install txtai from GitHub and set the defaultrole

```python
llm("Where is one place you'd go in Washington, DC?", defaultrole="user")
```

Now with the `Embeddings`, all of the supported models are documented here: https://docs.litellm.ai/docs/embedding/supported_embedding

I'm not exactly sure if Gemini models generate embeddings and/or if they are supported by LiteLLM though.

**Comment by igorlima** [comment]
Absolutely; thank you so much for the helpful hint! 🙏 🕺 

The resource shared in the previous comment was a game-changer. It successfully guided me through running the Gemini LLM pipeline and **generating embeddings** using Gemini. Here's a snippet of the code that worked seamlessly with the Gemini LLM model:

- <details><summary>Gemini <i>code snippet</i></summary>
  
  ```python
  import os
  os.environ["GEMINI_API_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
  
  from txtai import LLM
  # https://neuml.github.io/txtai/install/
  # https://neuml.github.io/txtai/pipeline/text/llm/#example
  # https://ai.google.dev/gemini-api/docs/models/gemini
  llm = LLM("gemini/gemini-pro", method="litellm")
  # print(llm("Where is one place you'd go in Washington, DC?", defaultrole="user"))
  print(llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}]))
  
  from txtai import Embeddings
  # https://neuml.github.io/txtai/embeddings/configuration/vectors/#method
  # https://docs.litellm.ai/docs/providers/gemini
  # https://github.com/BerriAI/litellm/tree/12c4e7e695edb07d403dd14fc768a736638bd3d1/litellm/llms/vertex_ai
  # https://github.com/BerriAI/litellm/blob/e19bb55e3b4c6a858b6e364302ebbf6633a51de5/model_prices_and_context_window.json#L2625
  embeddings = Embeddings(path="gemini/text-embedding-004", method="litellm")
  
  # works with a list, dataset or generator
  data = [
    "US tops 5 million confirmed virus cases",
    "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
    "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
    "The National Park Service warns against sacrificing slower friends in a bear attack",
    "Maine man wins $1M from $25 lottery ticket",
    "Make huge profits without work, earn up to $100,000 a day"
  ]
  
  # create an index for the list of text
  embeddings.index(data)
  print("%-20s %s" % ("Query", "Best Match"))
  print("-" * 50)
  # run an embeddings search for each query
  for query in ("feel good story", "climate change", "public health story", "war", "wildlife", "asia", "lucky", "dishonest junk"):
    # extract uid of first result
    # search result format: (uid, score)
    uid = embeddings.search(query, 1)[0][0]
    # print text
    print("%-20s %s" % (query, data[uid]))
  ```

  ![image](https://github.com/user-attachments/assets/d4d814d1-c62c-45e3-87ac-75e57e205d7a)
  </details>


Additionally, I’ve prepared four more samples for other models: VertexAI, Mistral, Cohere, and AWS Bedrock. Anyone searching for examples of these models can find them right here.

- <details><summary>other LLM models</summary>
  
  - <details><summary>VertexAI <i>code snippet</i></summary>

    ```python
    import os
    os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "application_default_credentials.json"
    
    import litellm
    # https://docs.litellm.ai/docs/embedding/supported_embedding#usage---embedding
    # https://docs.litellm.ai/docs/providers/vertex
    litellm.vertex_project = "hardy-device-38811" # Your Project ID
    litellm.vertex_location = "us-central1"  # proj location
    
    from txtai import LLM
    # https://neuml.github.io/txtai/install/
    # https://neuml.github.io/txtai/pipeline/text/llm/#example
    llm = LLM("vertex_ai/gemini-pro", method="litellm")
    
    # print(llm("Where is one place you'd go in Washington, DC?", defaultrole="user"))
    print(llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}]))
    
    from txtai import Embeddings
    # https://neuml.github.io/txtai/embeddings/configuration/vectors/#method
    # https://docs.litellm.ai/docs/providers/vertex
    # embeddings = Embeddings(path="text-embedding-004", method="litellm")
    embeddings = Embeddings(path="vertex_ai/text-embedding-004", method="litellm")
    
    # works with a list, dataset or generator
    data = [
      "US tops 5 million confirmed virus cases",
      "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
      "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
      "The National Park Service warns against sacrificing slower friends in a bear attack",
      "Maine man wins $1M from $25 lottery ticket",
      "Make huge profits without work, earn up to $100,000 a day"
    ]
    
    # create an index for the list of text
    embeddings.index(data)
    print("%-20s %s" % ("Query", "Best Match"))
    print("-" * 50)
    # run an embeddings search for each query
    for query in ("feel good story", "climate change", 
        "public health story", "war", "wildlife", "asia",
        "lucky", "dishonest junk"):
      # extract uid of first result
      # search result format: (uid, score)
      uid = embeddings.search(query, 1)[0][0]
      # print text
      print("%-20s %s" % (query, data[uid]))
    ```
    
    </details>
  - <details><summary>Mistral <i>code snippet</i></summary>

    ```python
    import os
    os.environ["MISTRAL_API_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
    
    from txtai import LLM
    # https://neuml.github.io/txtai/install/
    # https://neuml.github.io/txtai/pipeline/text/llm/#example
    llm = LLM("mistral/mistral-tiny", method="litellm")
    # print(llm("Where is one place you'd go in Washington, DC?", defaultrole="user"))
    print(llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}]))
    
    import litellm
    litellm.set_verbose=False
    
    from txtai import Embeddings
    # https://neuml.github.io/txtai/embeddings/configuration/vectors/#method
    # https://docs.litellm.ai/docs/providers/mistral
    embeddings = Embeddings(path="mistral/mistral-embed", method="litellm")
    
    # works with a list, dataset or generator
    data = [
      "US tops 5 million confirmed virus cases",
      "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
      "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
      "The National Park Service warns against sacrificing slower friends in a bear attack",
      "Maine man wins $1M from $25 lottery ticket",
      "Make huge profits without work, earn up to $100,000 a day"
    ]
    
    # create an index for the list of text
    embeddings.index(data)
    print("%-20s %s" % ("Query", "Best Match"))
    print("-" * 50)
    # run an embeddings search for each query
    for query in ("feel good story", "climate change", 
        "public health story", "war", "wildlife", "asia",
        "lucky", "dishonest junk"):
      # extract uid of first result
      # search result format: (uid, score)
      uid = embeddings.search(query, 1)[0][0]
      # print text
      print("%-20s %s" % (query, data[uid]))
    ```
    
    </details>
  - <details><summary>Cohere <i>code snippet</i></summary>

    ```python
    import os
    os.environ["COHERE_API_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
    
    from txtai import LLM
    # https://neuml.github.io/txtai/install/
    # https://neuml.github.io/txtai/pipeline/text/llm/#example
    llm = LLM("command-r", method="litellm")
    
    # print(llm("Where is one place you'd go in Washington, DC?", defaultrole="user"))
    print(llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}]))
    
    
    from txtai import Embeddings
    # https://neuml.github.io/txtai/embeddings/configuration/vectors/#method
    # https://docs.litellm.ai/docs/providers/cohere
    embeddings = Embeddings(path="cohere/embed-english-v3.0", method="litellm")
    
    # works with a list, dataset or generator
    data = [
      "US tops 5 million confirmed virus cases",
      "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
      "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
      "The National Park Service warns against sacrificing slower friends in a bear attack",
      "Maine man wins $1M from $25 lottery ticket",
      "Make huge profits without work, earn up to $100,000 a day"
    ]
    
    # create an index for the list of text
    embeddings.index(data)
    print("%-20s %s" % ("Query", "Best Match"))
    print("-" * 50)
    # run an embeddings search for each query
    for query in ("feel good story", "climate change", 
        "public health story", "war", "wildlife", "asia",
        "lucky", "dishonest junk"):
      # extract uid of first result
      # search result format: (uid, score)
      uid = embeddings.search(query, 1)[0][0]
      # print text
      print("%-20s %s" % (query, data[uid]))
    ```
    
    </details>
  - <details><summary>AWS Bedrock <i>code snippet</i></summary>

    ```python
    import os
    os.environ["AWS_ACCESS_KEY_ID"] = "xxxxxxxxxxxxxxxxxxxx"  # Access key
    os.environ["AWS_SECRET_ACCESS_KEY"] = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" # Secret access key
    os.environ["AWS_REGION_NAME"] = "us-west-2" # us-east-1, us-east-2, us-west-1, us-west-2
    
    from txtai import LLM
    # https://neuml.github.io/txtai/install/
    # https://neuml.github.io/txtai/pipeline/text/llm/#example
    llm = LLM("bedrock/amazon.titan-text-lite-v1", method="litellm")
    
    # print(llm("Where is one place you'd go in Washington, DC?", defaultrole="user"))
    print(llm([{"role": "user", "content": "Where is one place you'd go in Washington, DC?"}]))
    
    
    from txtai import Embeddings
    # https://neuml.github.io/txtai/embeddings/configuration/vectors/#method
    # https://docs.litellm.ai/docs/providers/bedrock
    embeddings = Embeddings(path="amazon.titan-embed-text-v1", method="litellm")
    
    # works with a list, dataset or generator
    data = [
      "US tops 5 million confirmed virus cases",
      "Canada's last fully intact ice shelf has suddenly collapsed, forming a Manhattan-sized iceberg",
      "Beijing mobilises invasion craft along coast as Taiwan tensions escalate",
      "The National Park Service warns against sacrificing slower friends in a bear attack",
      "Maine man wins $1M from $25 lottery ticket",
      "Make huge profits without work, earn up to $100,000 a day"
    ]
    
    # create an index for the list of text
    embeddings.index(data)
    print("%-20s %s" % ("Query", "Best Match"))
    print("-" * 50)
    # run an embeddings search for each query
    for query in ("feel good story", "climate change", 
        "public health story", "war", "wildlife", "asia",
        "lucky", "dishonest junk"):
      # extract uid of first result
      # search result format: (uid, score)
      uid = embeddings.search(query, 1)[0][0]
      # print text
      print("%-20s %s" % (query, data[uid]))
    ```
    
    </details>
  - <details><summary><code>requirements.txt</code></summary>

    ```env
    # python3 -m venv my-env
    # source my-env/bin/activate
    #
    # pip3 show txtai
    # pip3 index versions txtai
    # pip3 index versions txtai | grep -E "[(][0-9]+([.][0-9]+)+[)]"
    # pip3 list
    #
    # pip3 install --no-cache --upgrade-strategy eager -I txtai==8.1.0
    #
    # pip3 install -r requirements.txt
    #
    txtai==8.1.0
    model2vec==0.3.3
    pudb==2024.1.3
    
    # For macOS, there's a runtime dependency on `libomp`.
    # Run `brew install libomp` in this case.
    #
    # For Linux, there's a runtime dependency on `libsndfile`.
    # https://github.com/bastibe/python-soundfile#installation
    # https://github.com/libsndfile/libsndfile/issues/297
    # ```
    # apt update
    # apt search libsndfile
    # apt install libsndfile1
    # apt install libsndfile1-dev
    # ```
    txtai[pipeline]
    
    # to use Vertex AI
    # https://github.com/BerriAI/litellm/issues/5483
    google-cloud-aiplatform==1.75.0
    
    # to use AWS Bedrock
    # https://docs.litellm.ai/docs/providers/bedrock
    boto3==1.35.88
    ```
    </details>
  </details>

Once again, thank you for your support and guidance! 🙌 🎉

**Comment by davidmezzetti** [comment]
I'm glad this worked!

One minor thing, you shouldn't need `method='litellm'`  but it doesn't hurt either.

I appreciate you documenting how to do this with all the model endpoints you did, thank you!
